### DISCUSSION TITLE
classes/class_tcpserver

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#204] [User-contributed notes]
# classes/class_tcpserver

Inherits: RefCounted< Object A TCP server. Description: A TCP server. Listens to connections on a port and returns a StreamPeerTCP when it gets an incoming connection. Note: When exporting to Andro...

https://docs.godotengine.org/en/stable/classes/class_tcpserver.html

**Comment by aaron-bigadventurecrew** [comment]
Just had a good conversation with ChatGPT on a sample setup:
[https://pastebin.com/SYJpErW4](https://pastebin.com/SYJpErW4)

Not sure what's up with TCPServer's receipt of String data from a client's `put_data()` but switching to Variant cleaned things up nicely.

i.e. if client here is a StreamPeerTCP connected with a status of 2 to a TCPServer, I could see the data over the wire with Wireshark, but TCPServer always saw the raw data as [0, 0]:
`var incoming_data: PackedByteArray = client.get_data(bytes_to_read)`

But when I changed to using a Variant from the client and in the server it worked:
`var incoming_variant: Variant = client.get_var()`

I don't think there's a bug, but I could not sus out anything wrong with the initial attempt at using Strings, but Variant is probably better long-term anyways since you can then JSON/YAML it up. It was just weird because String or Variant, over the wire I saw valid packets getting [PSH, ACK] from the server:

0000   02 00 00 00 45 00 00 40 49 1c 40 00 80 06 00 00
0010   7f 00 00 01 7f 00 00 01 f1 bc a4 55 ae d5 66 d6
0020   61 b8 60 5d 50 18 27 f9 75 8f 00 00 04 00 00 00
0030   10 00 00 00 48 65 6c 6c 6f 20 57 6f 72 6c 64 21
0040   20 2d 20 39

¯\_(ツ)_/¯

**Reply by bamartin** [reply]
I had something similar, I was obviously getting a connection via StreamPeerTCP, but could only ever see [0,0] as a result. I found that I was using HTTPS on my loopback address in both my code and the URI redirect configuration in Google. So it was working, but I think I was getting encrypted traffic back that the TCP Peer couldn't understand. As soon as I changed it to use unsecure HTTP with my loopback address, I was able to see the appropriate response no problem.

**Comment by Varonex** [comment]
I have done a few TCP servers on Java & other languages, but for the love of me I fail to understand Godot's version of TCP server & sockets

Would it be possible to get an example of how they work? The UDPServer class had both a server & a client example so that we could understand

After some researches, I really realised that there are literally no posts thoroughly explaining what to do, which objects represent what, and I am not the only one to feel that way, as far as I'm concerned

**Comment by I2amthe1** [comment]
Simple TCP server example:

Replace contents in use_stream() with your desired behavior.
```gdscript
extends Node;

const DEFAULT_PORT: int = 19388;
var server: TCPServer;


func use_stream(stream: StreamPeerTCP):
    # Your Code Here
    pass;

func _ready():
    server = TCPServer.new();
    
    server.listen(DEFAULT_PORT);
func _process(_dt: float):
    while server.is_connection_available():
        var stream: StreamPeerTCP = server.take_connection();


        use_stream(stream);
```
